jujugui/static/gui/src/app/state/state.js

```diff
@@ -117,30 +149,192 @@ const State = class State {
   }
 
   /**
+    Stores the dispatchers that are to be called when the appropriate state
+    changes in the application.
+    @param {Array} dispatchers - An array of dispatchers in the format:
+      [['section', callback], ...]
+  */
+  register(dispatchers) {
+    const stored = this._dispatchers;
+    dispatchers.forEach(dispatcher => {
+      const record = [dispatcher[1], dispatcher[2]];
+      if (!stored[dispatcher[0]]) {
+        stored[dispatcher[0]] = [record];
+        return;
+      }
+      stored[dispatcher[0]].push(record);
+    });
+  }
+
+  /**
     Checks the current location and parses it, building the state, then
     executing the registered dispatchers.
+    @param {Array} nullKeys - A list of keys which we must run the 'cleanup'
+      dispaches on before dispatching the current state.
+    @param {Boolean} updateState - Whether this dispatch should update the
+      internal app state. This is typically only done when dispatch is called
+      manually. Internal calls to dispatch are done via changeState which
+      updates the state itself so it's not needed to do it here.
   */
-  dispatch() {
+  dispatch(nullKeys = [], updateState = true) {
     let error, state;
     ({error, state} = this.generateState(this.location.href));
-    this._appStateHistory.push(state);
     if (error !== null) {
-      return `unable to generate state: ${error}`;
+      error += `unable to generate state: ${error}`;
+      return {error, state};
+    }
+
+    /**
+      Extract out the state path for the dispatcher key.
+      @param {Object} state - The app state.
+    */
+    function extract(state) {
+      let allKeys = [];
+      function concat(state, keys = []) {
+        Object.keys(state).forEach(key => {
+          keys.push(key);
+          if (typeof state[key] === 'object' && state[key] !== null) {
+            concat(state[key], keys);
+          } else {
+            allKeys.push(keys.join('.'));
+          }
+          keys = [];
+        });
+      }
+      concat(state);
+      return allKeys;
     }
-    this._dispatch(state);
+    if (updateState) {
+      this._appStateHistory.push(state);
+    }
+    // First run all of the 'null state' dispatchers to clear out the old
+    // state representations.
+    nullKeys.forEach(key => this._dispatch(state, key, true));
+    // Then execute the 'all' dispatchers.
+    this._dispatch(state, '*');
+    // Extract and loop through the state keys to execute their dispatchers.
+    extract(state).forEach(key => {
+      this._dispatch(state, key);
```

Comment:
This part of the code has been rewritten to support the 'cleanup' step. But this doesn't because there is no 'all' state, it's just an artifact of the dispatch process here.